

##################
######################################################
##  This code calculates the promoter loading index using RNA polymerase II 
##  ChIP-seq and can be adapted for GRO-seq data as well
##  in this code the TSS region can be any length defined by the user as long as it is divsible by 2 and centered at TSS of gene(s)


# TSSR calculator ---------------------------------------------------------

#This function takes in a dataframe that has the name/unique identifier of a gene, chromosome, start and end co-ordinates 
#and generates a SAF format file to be used by Rsubread later.

TSSR.calculator<-function(myfile,strand.col,Start.col,End.col,chrom.col,Id.col, region.length) {
  ##region.length has to be a multiple of 2
  x<-myfile
  x$Start<-x[,Start.col] - round((region.length/2),0)
  x$Start[(which(x[,strand.col] == "-"))]<- x[(which(x[,strand.col] == "-")),End.col] - round((region.length/2),0)
  x$End<-x[,Start.col] + round((region.length/2),0)
  x$End[(which(x[,strand.col] == "-"))]<- x[(which(x[,strand.col] == "-")),End.col] + round((region.length/2),0)
  x$length<-region.length
  x<-x[,c(Id.col,chrom.col,Start.col,End.col,strand.col,(ncol(myfile)+1))]
  colnames(x)<-c("GeneID","Chr","Start","End","Strand","Length")
  x<-x[c(which(x$Start >= 0)),]
  return(x)
}



# Gene body region (GBR) calculator ---------------------------------------------------------

#This function takes in a dataframe that has the name/unique identifier of a gene, chromosome, start and end co-ordinates 
#and generates a SAF format file to be used by Rsubread later for the gene body region.

GBR.calculator<-function(myfile,strand.col,Start.col,End.col,chrom.col,Id.col, TSSR.length) {
  ##TSSR.length has to be a multiple of 2
  x<-myfile
  x$Start<-x[,Start.col] + (round((TSSR.length/2),0)+1)
  x$Start[(which(x[,strand.col] == "-"))]<- x[(which(x[,strand.col] == "-")),Start.col]
  x$End<-x[,End.col]
  x$End[(which(x[,strand.col] == "-"))]<- x[(which(x[,strand.col] == "-")),End.col] - (round((TSSR.length/2),0)+1)
  x$length<-(x[,End.col] - x[,Start.col] - (round((TSSR.length/2),0)+1))
  #removing genes or transcripts whose size is smaller than the range used
  x<-x[c(which(x$length >0 )),]
  x<-x[,c(Id.col,chrom.col,Start.col,End.col,strand.col,(ncol(myfile)+1))]
  colnames(x)<-c("GeneID","Chr","Start","End","Strand","Length")
  return(x)
}


# Region_count_generator --------------------------------------------------
#This function

Region_count_generator <-function(bam_files, SAF_file, out.dir, threads,
                                Paired=TRUE, #if experiment is Paired-end
                                type, #Depends on the experiment can be either "ChIP" for ChIP-seq 
                              # or "GRO" for GRO-seq or PRO-seq, experiments
                                strand # Only applicable if experiment is GRO or Pro-seq
                              # This can be  0 for unstranded library preperation,
                              # 1 is for stranded (the R1 read is in the same direction as the RNA fragment)
                              # 2 is for reversely stranded library (such as the Illumina Tru-seq library preperation kit)
                              ) {
  require(Rsubread)
  if (type == "ChIP") {
  #counting reads in these regions
    if (Paired == TRUE){
      y<-featureCounts(bam_files, # a list of bam files
                       annot.ext=SAF_file, #SAFfile generated by TSSR.calculator
                       nthreads=threads,#number of threads/processors used
                       countMultiMappingReads=TRUE, #Counting reads that map to multiple features (can be adjacent etc.)
                       ignoreDup=TRUE, # ignores PCR duplicates
                       allowMultiOverlap=TRUE, #allows reads that map to multiple overlapping features to be counted for both features
                       isPairedEnd=TRUE, #Depends if youe experiment is single-end or paired-end
                       strandSpecific=0, #if the library preperation is stranded (Only applicable for Gro-seq experiments)
                       )
    } else if (Paired == FALSE){
      y<-featureCounts(bam_files, # a list of bam files
                       annot.ext=SAF_file, #SAF file generated by TSSR.calculator
                       nthreads=threads,#number of threads/processors used
                       countMultiMappingReads=TRUE, #Counting reads that map to multiple features (can be adjacent etc.)
                       ignoreDup=TRUE, # ignores PCR duplicates
                       allowMultiOverlap=TRUE, #allows reads that map to multiple overlapping features to be counted for both features
                       isPairedEnd=FALSE, #Depends if youe experiment is single-end or paired-end
                       strandSpecific=0, #if the library preperation is stranded (Only applicable for Gro-seq experiments)
                       ignoreDup=TRUE)
    } else { stop("Sequencing type not correctly specified\nYou have to specify if Paired=True or FALSE")}
  } else if (type=="GRO"){
    if (strand == 0|strand==1|strand==2){
      if (Paired == TRUE){
        y<-featureCounts(bam_files, # a list of bam files
                         annot.ext=SAF_file, #SAFfile generated by TSSR.calculator
                         nthreads=threads,#number of threads/processors used
                         countMultiMappingReads=TRUE, #Counting reads that map to multiple features (can be adjacent etc.)
                         ignoreDup=TRUE, # ignores PCR duplicates
                         allowMultiOverlap=TRUE, #allows reads that map to multiple overlapping features to be counted for both features
                         isPairedEnd=TRUE, #Depends if youe experiment is single-end or paired-end
                         strandSpecific=strand, #if the library preperation is stranded (Only applicable for Gro-seq experiments)
        )
      } else if (Paired == FALSE) {
        y<-featureCounts(bam_files, # a list of bam files
                         annot.ext=SAF_file, #SAFfile generated by TSSR.calculator
                         nthreads=threads,#number of threads/processors used
                         countMultiMappingReads=TRUE, #Counting reads that map to multiple features (can be adjacent etc.)
                         ignoreDup=TRUE, # ignores PCR duplicates
                         allowMultiOverlap=TRUE, #allows reads that map to multiple overlapping features to be counted for both features
                         isPairedEnd=FALSE, #Depends if youe experiment is single-end or paired-end
                         strandSpecific=strand, #if the library preperation is stranded (Only applicable for Gro-seq experiments)
                         ignoreDup=TRUE)
      } else { stop("Sequencing type not correctly specified\nYou have to specify if Paired=True or FALSE")}
    } else { stop("library strandedness not correctly specified\nYou have to specify if strand=0, or strand=1 or strand=2")}
  } else { stop('Experiment type not specified\nYou have to indicate if type="ChIP", or type="GRO"')}
  y<-as.data.frame(y$counts)
  y<-merge(y,SAF_file)
  return(y)
  gc()
  }



      
